src/reader.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "reader.h"
+
+#include "fileops.h"
+#include "blob.h"
+
+#include "git2/tree.h"
+#include "git2/blob.h"
+#include "git2/index.h"
+#include "git2/repository.h"
+
+/* tree reader */
+
+typedef struct {
+	git_reader reader;
+	git_tree *tree;
+} tree_reader;
+
+static int tree_reader_read(
+	git_buf *out,
+	git_oid *out_id,
+	git_reader *_reader,
+	const char *filename)
+{
+	tree_reader *reader = (tree_reader *)_reader;
+	git_tree_entry *tree_entry = NULL;
+	git_blob *blob = NULL;
+	int error;
+
+	if ((error = git_tree_entry_bypath(&tree_entry,
+		reader->tree, filename)) < 0 ||
+		(error = git_blob_lookup(&blob,
+		git_tree_owner(reader->tree), git_tree_entry_id(tree_entry))) < 0)
+		goto done;
+
+	git_buf_clear(out);
+	git_buf_put(out, git_blob_rawcontent(blob), git_blob_rawsize(blob));
+
+	if (git_buf_oom(out))
+		error = -1;
+
+	if (out_id)
+		git_oid_cpy(out_id, git_tree_entry_id(tree_entry));
+
+done:
+	git_blob_free(blob);
+	git_tree_entry_free(tree_entry);
+	return error;
+}
+
+static void tree_reader_free(git_reader *_reader)
+{
+	GIT_UNUSED(_reader);
+}
+
+int git_reader_for_tree(git_reader **out, git_tree *tree)
+{
+	tree_reader *reader;
+
+	reader = git__calloc(1, sizeof(tree_reader));
+	GITERR_CHECK_ALLOC(reader);
+
+	reader->reader.read = tree_reader_read;
+	reader->reader.free = tree_reader_free;
+	reader->tree = tree;
+
+	*out = (git_reader *)reader;
+	return 0;
+}
+
+/* workdir reader */
+
+typedef struct {
+	git_reader reader;
+	git_repository *repo;
+	git_index *index;
+} workdir_reader;
+
+static int workdir_reader_read(
+	git_buf *out,
+	git_oid *out_id,
+	git_reader *_reader,
+	const char *filename)
+{
+	workdir_reader *reader = (workdir_reader *)_reader;
+	git_buf path = GIT_BUF_INIT;
+	git_filter_list *filters = NULL;
+	const git_index_entry *idx_entry;
+	git_oid id;
+	int error;
+
+	if ((error = git_buf_joinpath(&path,
+		git_repository_workdir(reader->repo), filename)) < 0)
+		goto done;
+
+	if ((error = git_filter_list_load(&filters, reader->repo, NULL, filename,
+		GIT_FILTER_TO_ODB, GIT_FILTER_DEFAULT)) < 0)
+		goto done;
+
+	if ((error = git_filter_list_apply_to_file(out,
+	    filters, reader->repo, path.ptr)) < 0)
+		goto done;
+
+	if (out_id || reader->index) {
+		if ((error = git_odb_hash(&id, out->ptr, out->size, GIT_OBJ_BLOB)) < 0)
+			goto done;
+	}
+
+	if (reader->index) {
+		if (!(idx_entry = git_index_get_bypath(reader->index, filename, 0)) ||
+		    !git_oid_equal(&id, &idx_entry->id)) {
+			error = GIT_READER_MISMATCH;
+			goto done;
+		}
```

Comment:
Ah, you need the cleaned file to verify it's got the same content as the index. I still wonder about whether it's the right thing to do in the context of git-apply(1)

Reply:
I'm copying git's behavior here.  It runs the workdir -> odb filter as well.  This is tough to see with (for example) line ending changes, since line endings are already a mess and patch application tries to deal with them.

However, it's readily evident if you put a text file into LFS.  Here I'll add a simple text file to the repository and apply a patch:

```
> git add bar.txt

>git commit -m"bar"
[master (root-commit) ebc14bf] bar
 1 file changed, 16 insertions(+)
 create mode 100644 bar.txt

>type bar.txt
0. Hello, this file has CR/LF line endings.
1. Hello, this file has CR/LF line endings.
2. Hello, this file has CR/LF line endings.
3. Hello, this file has CR/LF line endings.
4. Hello, this file has CR/LF line endings.
5. Hello, this file has CR/LF line endings.
6. Hello, this file has CR/LF line endings.
7. Hello, this file has CR/LF line endings.
8. Hello, this file has CR/LF line endings.
9. Hello, this file has CR/LF line endings.
a. Hello, this file has CR/LF line endings.
b. Hello, this file has CR/LF line endings.
c. Hello, this file has CR/LF line endings.
d. Hello, this file has CR/LF line endings.
e. Hello, this file has CR/LF line endings.
f. Hello, this file has CR/LF line endings.

> git apply < patch

> type bar.txt
0. Hello, this file has CR/LF line endings.
1. Hello, this file has CR/LF line endings.
2. Hello, this file has CR/LF line endings.
3. Hello, this file has CR/LF line endings.
4. Here's a change on the fifth line.
5. Hello, this file has CR/LF line endings.
6. Hello, this file has CR/LF line endings.
7. Hello, this file has CR/LF line endings.
8. Hello, this file has CR/LF line endings.
9. Hello, this file has CR/LF line endings.
a. Hello, this file has CR/LF line endings.
b. Here's a change on another line.
c. Hello, this file has CR/LF line endings.
d. Hello, this file has CR/LF line endings.
e. Hello, this file has CR/LF line endings.
f. Hello, this file has CR/LF line endings.
```

Now let's add the file as an LFS resource and try to do the same thing:

```
> git reset --hard HEAD
HEAD is now at ebc14bf bar

> git lfs install
Updated git hooks.
Git LFS initialized.

> git lfs track bar.txt
Tracking "bar.txt"

> git add bar.txt

> git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   bar.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        .gitattributes
        patch

>git commit -mbar
[master 756794c] bar
 1 file changed, 3 insertions(+), 16 deletions(-)
 rewrite bar.txt (100%)

> git apply < patch
error: patch failed: bar.txt:2
error: bar.txt: patch does not apply
```

Which is, of course, because it's applying the filter.  Which suggests that we could actually patch the LFS pointer file.  Which, indeed, we can:

```
>git ls-files --stage
100644 976ef879085369af40ec959843624d6c5b0f5024 0       bar.txt

>git cat-file blob 976ef
version https://git-lfs.github.com/spec/v1
oid sha256:cad943fa9780d676a5236c9fe503d8f63b2115cf3786d0884bc8c07a84aa9d92
size 720

> type patch
diff --git a/bar.txt b/bar.txt
index 976ef87..d2c468a 100644
--- a/bar.txt
+++ b/bar.txt
@@ -1,3 +1,4 @@
 version https://git-lfs.github.com/spec/v1
 oid sha256:cad943fa9780d676a5236c9fe503d8f63b2115cf3786d0884bc8c07a84aa9d92
 size 720
+Yes, of course it's patching the LFS pointer file.

> git apply < patch
Encountered 1 file(s) that should have been pointers, but weren't:
        bar.txt

> type bar.txt
version https://git-lfs.github.com/spec/v1
oid sha256:cad943fa9780d676a5236c9fe503d8f63b2115cf3786d0884bc8c07a84aa9d92
size 720
Yes, of course it's patching the LFS pointer file.
```

Anyway.  I don't really _like_ this design decision.  It seems to me that you would want to patch the filtered content, not the ODB content.  I think that it was probably designed to deal with line ending, like oh so much else.  But whether I like this decision or not, like oh so many things, it's not really ours to make.

Reply:
Added a bit of documentation here, as well.